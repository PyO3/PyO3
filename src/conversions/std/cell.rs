use std::cell::Cell;

use crate::{
    conversion::IntoPyObject, types::any::PyAnyMethods, Bound, FromPyObject, IntoPy, PyAny,
    PyObject, PyResult, Python, ToPyObject,
};

impl<T: Copy + ToPyObject> ToPyObject for Cell<T> {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        self.get().to_object(py)
    }
}

impl<T: Copy + IntoPy<PyObject>> IntoPy<PyObject> for Cell<T> {
    fn into_py(self, py: Python<'_>) -> PyObject {
        self.get().into_py(py)
    }
}

impl<'py, Target, T: Copy + IntoPyObject<'py, Target>> IntoPyObject<'py, Target> for Cell<T> {
    type Error = T::Error;

    fn into_pyobj(self, py: Python<'py>) -> Result<Bound<'py, Target>, Self::Error> {
        self.get().into_pyobj(py)
    }
}

impl<'py, T: FromPyObject<'py>> FromPyObject<'py> for Cell<T> {
    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
        ob.extract().map(Cell::new)
    }
}
