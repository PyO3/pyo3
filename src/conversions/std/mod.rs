mod array;
mod cell;
mod ipaddr;
mod map;
mod num;
mod option;
mod osstr;
mod path;
mod set;
mod slice;
mod string;
mod time;
mod vec;
