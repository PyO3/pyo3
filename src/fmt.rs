//! This module provides the `PyUnicodeWriter` struct, which is a utility for efficiently
//! constructing Python strings using Rust's `fmt::Write` trait.
//! It allows for incremental string construction, without the need for repeated allocations, and
//! is particularly useful for building strings in a performance-sensitive context.
#[cfg(not(any(Py_LIMITED_API, PyPy)))]
use {
    crate::ffi::compat::{
        PyUnicodeWriter_Create, PyUnicodeWriter_Discard, PyUnicodeWriter_Finish,
        PyUnicodeWriter_WriteChar, PyUnicodeWriter_WriteUTF8,
    },
    crate::ffi_ptr_ext::FfiPtrExt,
    crate::impl_::callback::WrappingCastTo,
    crate::py_result_ext::PyResultExt,
    crate::types::PyString,
    crate::IntoPyObject,
    crate::{ffi, Bound, PyErr, PyResult, Python},
    std::fmt,
    std::mem::ManuallyDrop,
    std::ptr::NonNull,
};

/// This is like the `format!` macro, but it returns a `PyString` instead of a `String`.
#[macro_export]
macro_rules! py_format {
    ($py: expr, $($arg:tt)*) => {
        $crate::types::PyString::from_fmt($py, format_args!($($arg)*))
    }
}

#[cfg(not(any(Py_LIMITED_API, PyPy)))]
/// The `PyUnicodeWriter` is a utility for efficiently constructing Python strings
pub struct PyUnicodeWriter {
    writer: NonNull<ffi::PyUnicodeWriter>,
    last_error: Option<PyErr>,
}

#[cfg(not(any(Py_LIMITED_API, PyPy)))]
impl PyUnicodeWriter {
    /// Creates a new `PyUnicodeWriter`.
    pub fn new(py: Python<'_>) -> PyResult<Self> {
        Self::with_capacity(py, 0)
    }

    /// Creates a new `PyUnicodeWriter` with the specified initial capacity.
    pub fn with_capacity(py: Python<'_>, capacity: usize) -> PyResult<Self> {
        match NonNull::new(unsafe { PyUnicodeWriter_Create(capacity.wrapping_cast()) }) {
            Some(ptr) => Ok(PyUnicodeWriter {
                writer: ptr,
                last_error: None,
            }),
            None => Err(PyErr::fetch(py)),
        }
    }

    /// Consumes the `PyUnicodeWriter` and returns a `Bound<PyString>` containing the constructed string.
    pub fn into_py_string(mut self, py: Python<'_>) -> PyResult<Bound<'_, PyString>> {
        if let Some(error) = self.take_error() {
            Err(error)
        } else {
            unsafe {
                PyUnicodeWriter_Finish(ManuallyDrop::new(self).as_ptr())
                    .assume_owned_or_err(py)
                    .cast_into_unchecked()
            }
        }
    }

    /// When fmt::Write returned an error, this function can be used to retrieve the last error that occurred.
    pub fn take_error(&mut self) -> Option<PyErr> {
        self.last_error.take()
    }

    fn as_ptr(&self) -> *mut ffi::PyUnicodeWriter {
        self.writer.as_ptr()
    }

    fn set_error(&mut self) {
        Python::attach(|py| {
            self.last_error = Some(PyErr::fetch(py));
        })
    }
}

#[cfg(not(any(Py_LIMITED_API, PyPy)))]
impl fmt::Write for PyUnicodeWriter {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        let result = unsafe {
            PyUnicodeWriter_WriteUTF8(self.as_ptr(), s.as_ptr().cast(), s.len() as isize)
        };
        if result < 0 {
            self.set_error();
            Err(fmt::Error)
        } else {
            Ok(())
        }
    }

    fn write_char(&mut self, c: char) -> fmt::Result {
        let result = unsafe { PyUnicodeWriter_WriteChar(self.as_ptr(), c as u32) };
        if result < 0 {
            self.set_error();
            Err(fmt::Error)
        } else {
            Ok(())
        }
    }
}

#[cfg(not(any(Py_LIMITED_API, PyPy)))]
impl Drop for PyUnicodeWriter {
    fn drop(&mut self) {
        unsafe {
            PyUnicodeWriter_Discard(self.as_ptr());
        }
    }
}

#[cfg(not(any(Py_LIMITED_API, PyPy)))]
impl<'py> IntoPyObject<'py> for PyUnicodeWriter {
    type Target = PyString;
    type Output = Bound<'py, Self::Target>;
    type Error = PyErr;

    fn into_pyobject(self, py: Python<'py>) -> PyResult<Bound<'py, PyString>> {
        self.into_py_string(py)
    }
}

#[cfg(test)]
mod tests {
    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
    use super::*;
    use crate::types::PyStringMethods;
    use crate::{IntoPyObject, Python};

    #[test]
    #[allow(clippy::write_literal)]
    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
    fn unicode_writer_test() {
        use std::fmt::Write;
        Python::attach(|py| {
            let mut writer = PyUnicodeWriter::new(py).unwrap();
            write!(writer, "Hello {}!", "world").unwrap();
            writer.write_char('ðŸ˜Ž').unwrap();
            let result = writer.into_py_string(py).unwrap();
            assert_eq!(result.to_string(), "Hello world!ðŸ˜Ž");
        });
    }

    #[test]
    fn test_pystring_from_fmt() {
        Python::attach(|py| {
            py_format!(py, "Hello {}!", "world").unwrap();
        });
    }

    #[test]
    fn test_complex_format() {
        Python::attach(|py| {
            let complex_value = (42, "foo", [0; 0]).into_pyobject(py).unwrap();
            let py_string = py_format!(py, "This is some complex value: {complex_value}").unwrap();
            let actual = py_string.to_cow().unwrap();
            let expected = "This is some complex value: (42, 'foo', [])";
            assert_eq!(actual, expected);
        });
    }
}
