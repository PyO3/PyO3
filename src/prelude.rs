//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
