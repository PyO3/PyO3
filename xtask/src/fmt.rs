use crate::cli;
use std::process::Command;

pub fn run() -> anyhow::Result<()> {
    cli::run(
        Command::new("cargo")
            .arg("fmt")
            .arg("--all")
            .arg("--")
            .arg("--check"),
    )?;

    cli::run(Command::new("black").arg(".").arg("--check"))?;

    Ok(())
}
