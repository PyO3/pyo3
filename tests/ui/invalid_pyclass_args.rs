use pyo3::prelude::*;
use std::fmt::{Display, Formatter};

#[pyclass(extend=pyo3::types::PyDict)]
struct TypoIntheKey {}

#[pyclass(extends = "PyDict")]
struct InvalidExtends {}

#[pyclass(name = m::MyClass)]
struct InvalidName {}

#[pyclass(name = "Custom Name")]
struct InvalidName2 {}

#[pyclass(name = CustomName)]
struct DeprecatedName {}

#[pyclass(rename_all = camelCase)]
struct InvalidRenamingRule {}

#[pyclass(rename_all = "Camel-Case")]
struct InvalidRenamingRule2 {}

#[pyclass(module = my_module)]
struct InvalidModule {}

#[pyclass(weakrev)]
struct InvalidArg {}

#[pyclass(mapping, sequence)]
struct CannotBeMappingAndSequence {}

#[pyclass(eq)]
struct EqOptRequiresEq {}

#[pyclass(eq)]
#[derive(PartialEq)]
struct EqOptAndManualRichCmp {}

#[pymethods]
impl EqOptAndManualRichCmp {
    fn __richcmp__(
        &self,
        _py: Python,
        _other: Bound<'_, PyAny>,
        _op: pyo3::pyclass::CompareOp,
    ) -> PyResult<PyObject> {
        todo!()
    }
}

#[pyclass(eq_int)]
struct NoEqInt {}

#[pyclass(frozen, eq, hash)]
#[derive(PartialEq)]
struct HashOptRequiresHash;

#[pyclass(hash)]
#[derive(Hash)]
struct HashWithoutFrozenAndEq;

#[pyclass(frozen, eq, hash)]
#[derive(PartialEq, Hash)]
struct HashOptAndManualHash {}

#[pymethods]
impl HashOptAndManualHash {
    fn __hash__(&self) -> u64 {
        todo!()
    }
}

#[pyclass(ord)]
struct InvalidOrderedStruct {
    inner: i32,
}

#[pyclass]
struct MultipleErrors {
    #[pyo3(foo)]
    #[pyo3(blah)]
    x: i32,
    #[pyo3(pop)]
    y: i32,
}

#[pyclass(str)]
struct StrOptAndManualStr {}

impl Display for StrOptAndManualStr {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}

#[pymethods]
impl StrOptAndManualStr {
    fn __str__(&self) -> String {
        todo!()
    }
}

#[pyclass(str = "{")]
#[derive(PartialEq)]
struct Coord(u32, u32, u32);

#[pyclass(str = "{$}")]
#[derive(PartialEq)]
struct Coord2(u32, u32, u32);

#[pyclass(str = "X: {aaaa}, Y: {y}, Z: {z}")]
#[derive(PartialEq, Eq, Clone, PartialOrd)]
pub struct Point {
    x: i32,
    y: i32,
    z: i32,
}

#[pyclass(str = "X: {x}, Y: {y}}}, Z: {zzz}")]
#[derive(PartialEq, Eq, Clone, PartialOrd)]
pub struct Point2 {
    x: i32,
    y: i32,
    z: i32,
}

#[pyclass(str = "{0}, {162543}, {2}")]
#[derive(PartialEq)]
struct Coord3(u32, u32, u32);

#[pyclass(name = "aaa", str = "unsafe: {unsafe_variable}")]
struct StructRenamingWithStrFormatter {
    #[pyo3(name = "unsafe", get, set)]
    unsafe_variable: usize,
}

#[pyclass(name = "aaa", str = "unsafe: {unsafe_variable}")]
struct StructRenamingWithStrFormatter2 {
    unsafe_variable: usize,
}

#[pyclass(str = "unsafe: {unsafe_variable}")]
struct StructRenamingWithStrFormatter3 {
    #[pyo3(name = "unsafe", get, set)]
    unsafe_variable: usize,
}

#[pyclass(rename_all = "SCREAMING_SNAKE_CASE", str = "{a_a}, {b_b}, {c_d_e}")]
struct RenameAllVariantsStruct {
    a_a: u32,
    b_b: u32,
    c_d_e: String,
}

#[pyclass(str = "{:?}")]
#[derive(Debug)]
struct StructWithNoMember {
    a: String,
    b: String,
}

#[pyclass(str = "{}")]
#[derive(Debug)]
struct StructWithNoMember2 {
    a: String,
    b: String,
}

#[pyclass(eq, str = "Stuff...")]
#[derive(Debug, PartialEq)]
pub enum MyEnumInvalidStrFmt {
    Variant,
    OtherVariant,
}

impl Display for MyEnumInvalidStrFmt {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

fn main() {}
