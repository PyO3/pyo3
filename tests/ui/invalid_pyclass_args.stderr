error: expected one of: `crate`, `dict`, `eq`, `eq_int`, `extends`, `freelist`, `frozen`, `get_all`, `hash`, `mapping`, `module`, `name`, `ord`, `opaque`, `rename_all`, `sequence`, `set_all`, `str`, `subclass`, `unsendable`, `weakref`
 --> tests/ui/invalid_pyclass_args.rs:4:11
  |
4 | #[pyclass(extend=pyo3::types::PyDict)]
  |           ^^^^^^

error: expected identifier
 --> tests/ui/invalid_pyclass_args.rs:7:21
  |
7 | #[pyclass(extends = "PyDict")]
  |                     ^^^^^^^^

error: expected string literal
  --> tests/ui/invalid_pyclass_args.rs:10:18
   |
10 | #[pyclass(name = m::MyClass)]
   |                  ^

error: expected a single identifier in double quotes
  --> tests/ui/invalid_pyclass_args.rs:13:18
   |
13 | #[pyclass(name = "Custom Name")]
   |                  ^^^^^^^^^^^^^

error: expected string literal
  --> tests/ui/invalid_pyclass_args.rs:16:18
   |
16 | #[pyclass(name = CustomName)]
   |                  ^^^^^^^^^^

error: expected string literal
  --> tests/ui/invalid_pyclass_args.rs:19:24
   |
19 | #[pyclass(rename_all = camelCase)]
   |                        ^^^^^^^^^

error: expected a valid renaming rule, possible values are: "camelCase", "kebab-case", "lowercase", "PascalCase", "SCREAMING-KEBAB-CASE", "SCREAMING_SNAKE_CASE", "snake_case", "UPPERCASE"
  --> tests/ui/invalid_pyclass_args.rs:22:24
   |
22 | #[pyclass(rename_all = "Camel-Case")]
   |                        ^^^^^^^^^^^^

error: expected string literal
  --> tests/ui/invalid_pyclass_args.rs:25:20
   |
25 | #[pyclass(module = my_module)]
   |                    ^^^^^^^^^

error: expected one of: `crate`, `dict`, `eq`, `eq_int`, `extends`, `freelist`, `frozen`, `get_all`, `hash`, `mapping`, `module`, `name`, `ord`, `opaque`, `rename_all`, `sequence`, `set_all`, `str`, `subclass`, `unsendable`, `weakref`
  --> tests/ui/invalid_pyclass_args.rs:28:11
   |
28 | #[pyclass(weakrev)]
   |           ^^^^^^^

error: a `#[pyclass]` cannot be both a `mapping` and a `sequence`
  --> tests/ui/invalid_pyclass_args.rs:32:8
   |
32 | struct CannotBeMappingAndSequence {}
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `eq_int` can only be used on simple enums.
  --> tests/ui/invalid_pyclass_args.rs:53:11
   |
53 | #[pyclass(eq_int)]
   |           ^^^^^^

error: The `hash` option requires the `frozen` option.
  --> tests/ui/invalid_pyclass_args.rs:60:11
   |
60 | #[pyclass(hash)]
   |           ^^^^

error: The `hash` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_args.rs:60:11
   |
60 | #[pyclass(hash)]
   |           ^^^^

error: The `ord` option requires the `eq` option.
  --> tests/ui/invalid_pyclass_args.rs:75:11
   |
75 | #[pyclass(ord)]
   |           ^^^

error: expected one of: `get`, `set`, `name`
  --> tests/ui/invalid_pyclass_args.rs:82:12
   |
82 |     #[pyo3(foo)]
   |            ^^^

error: expected one of: `get`, `set`, `name`
  --> tests/ui/invalid_pyclass_args.rs:83:12
   |
83 |     #[pyo3(blah)]
   |            ^^^^

error: expected one of: `get`, `set`, `name`
  --> tests/ui/invalid_pyclass_args.rs:85:12
   |
85 |     #[pyo3(pop)]
   |            ^^^

error: invalid format string: expected `'}'` but string was terminated
   --> tests/ui/invalid_pyclass_args.rs:105:19
    |
105 | #[pyclass(str = "{")]
    |                  -^ expected `'}'` in format string
    |                  |
    |                  because of this opening brace
    |
    = note: if you intended to print `{`, you can escape it using `{{`

error: invalid format string: expected `'}'`, found `'$'`
   --> tests/ui/invalid_pyclass_args.rs:109:19
    |
109 | #[pyclass(str = "{$}")]
    |                  -^ expected `'}'` in format string
    |                  |
    |                  because of this opening brace
    |
    = note: if you intended to print `{`, you can escape it using `{{`

error: The format string syntax is incompatible with any renaming via `name` or `rename_all`
   --> tests/ui/invalid_pyclass_args.rs:133:31
    |
133 | #[pyclass(name = "aaa", str = "unsafe: {unsafe_variable}")]
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The format string syntax is incompatible with any renaming via `name` or `rename_all`
   --> tests/ui/invalid_pyclass_args.rs:139:31
    |
139 | #[pyclass(name = "aaa", str = "unsafe: {unsafe_variable}")]
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The format string syntax is incompatible with any renaming via `name` or `rename_all`
   --> tests/ui/invalid_pyclass_args.rs:144:17
    |
144 | #[pyclass(str = "unsafe: {unsafe_variable}")]
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: The format string syntax is incompatible with any renaming via `name` or `rename_all`
   --> tests/ui/invalid_pyclass_args.rs:150:54
    |
150 | #[pyclass(rename_all = "SCREAMING_SNAKE_CASE", str = "{a_a}, {b_b}, {c_d_e}")]
    |                                                      ^^^^^^^^^^^^^^^^^^^^^^^

error: No member found, you must provide a named or positionally specified member.
   --> tests/ui/invalid_pyclass_args.rs:157:17
    |
157 | #[pyclass(str = "{:?}")]
    |                 ^^^^^^

error: No member found, you must provide a named or positionally specified member.
   --> tests/ui/invalid_pyclass_args.rs:164:17
    |
164 | #[pyclass(str = "{}")]
    |                 ^^^^

error: The format string syntax cannot be used with enums
   --> tests/ui/invalid_pyclass_args.rs:171:21
    |
171 | #[pyclass(eq, str = "Stuff...")]
    |                     ^^^^^^^^^^

error[E0592]: duplicate definitions with name `__pymethod___richcmp____`
  --> tests/ui/invalid_pyclass_args.rs:37:1
   |
37 | #[pyclass(eq)]
   | ^^^^^^^^^^^^^^ duplicate definitions for `__pymethod___richcmp____`
...
41 | #[pymethods]
   | ------------ other definition for `__pymethod___richcmp____`
   |
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0592]: duplicate definitions with name `__pymethod___hash____`
  --> tests/ui/invalid_pyclass_args.rs:64:1
   |
64 | #[pyclass(frozen, eq, hash)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `__pymethod___hash____`
...
68 | #[pymethods]
   | ------------ other definition for `__pymethod___hash____`
   |
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0592]: duplicate definitions with name `__pymethod___str____`
  --> tests/ui/invalid_pyclass_args.rs:89:1
   |
89 | #[pyclass(str)]
   | ^^^^^^^^^^^^^^^ duplicate definitions for `__pymethod___str____`
...
98 | #[pymethods]
   | ------------ other definition for `__pymethod___str____`
   |
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0369]: binary operation `==` cannot be applied to type `&EqOptRequiresEq`
  --> tests/ui/invalid_pyclass_args.rs:34:11
   |
34 | #[pyclass(eq)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `EqOptRequiresEq`
  --> tests/ui/invalid_pyclass_args.rs:35:1
   |
35 | struct EqOptRequiresEq {}
   | ^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `EqOptRequiresEq` with `#[derive(PartialEq)]`
   |
35 + #[derive(PartialEq)]
36 | struct EqOptRequiresEq {}
   |

error[E0369]: binary operation `!=` cannot be applied to type `&EqOptRequiresEq`
  --> tests/ui/invalid_pyclass_args.rs:34:11
   |
34 | #[pyclass(eq)]
   |           ^^
   |
note: an implementation of `PartialEq` might be missing for `EqOptRequiresEq`
  --> tests/ui/invalid_pyclass_args.rs:35:1
   |
35 | struct EqOptRequiresEq {}
   | ^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `EqOptRequiresEq` with `#[derive(PartialEq)]`
   |
35 + #[derive(PartialEq)]
36 | struct EqOptRequiresEq {}
   |

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:37:1
   |
37 | #[pyclass(eq)]
   | ^^^^^^^^^^^^^^ multiple `__pymethod___richcmp____` found
   |
note: candidate #1 is defined in an impl for the type `EqOptAndManualRichCmp`
  --> tests/ui/invalid_pyclass_args.rs:37:1
   |
37 | #[pyclass(eq)]
   | ^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `EqOptAndManualRichCmp`
  --> tests/ui/invalid_pyclass_args.rs:41:1
   |
41 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pyclass` which comes from the expansion of the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:41:1
   |
41 | #[pymethods]
   | ^^^^^^^^^^^^ multiple `__pymethod___richcmp____` found
   |
note: candidate #1 is defined in an impl for the type `EqOptAndManualRichCmp`
  --> tests/ui/invalid_pyclass_args.rs:37:1
   |
37 | #[pyclass(eq)]
   | ^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `EqOptAndManualRichCmp`
  --> tests/ui/invalid_pyclass_args.rs:41:1
   |
41 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `HashOptRequiresHash: Hash` is not satisfied
  --> tests/ui/invalid_pyclass_args.rs:56:23
   |
56 | #[pyclass(frozen, eq, hash)]
   |                       ^^^^ the trait `Hash` is not implemented for `HashOptRequiresHash`
   |
help: consider annotating `HashOptRequiresHash` with `#[derive(Hash)]`
   |
58 + #[derive(Hash)]
59 | struct HashOptRequiresHash;
   |

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:64:1
   |
64 | #[pyclass(frozen, eq, hash)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ multiple `__pymethod___hash____` found
   |
note: candidate #1 is defined in an impl for the type `HashOptAndManualHash`
  --> tests/ui/invalid_pyclass_args.rs:64:1
   |
64 | #[pyclass(frozen, eq, hash)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `HashOptAndManualHash`
  --> tests/ui/invalid_pyclass_args.rs:68:1
   |
68 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pyclass` which comes from the expansion of the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:68:1
   |
68 | #[pymethods]
   | ^^^^^^^^^^^^ multiple `__pymethod___hash____` found
   |
note: candidate #1 is defined in an impl for the type `HashOptAndManualHash`
  --> tests/ui/invalid_pyclass_args.rs:64:1
   |
64 | #[pyclass(frozen, eq, hash)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `HashOptAndManualHash`
  --> tests/ui/invalid_pyclass_args.rs:68:1
   |
68 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:89:1
   |
89 | #[pyclass(str)]
   | ^^^^^^^^^^^^^^^ multiple `__pymethod___str____` found
   |
note: candidate #1 is defined in an impl for the type `StrOptAndManualStr`
  --> tests/ui/invalid_pyclass_args.rs:89:1
   |
89 | #[pyclass(str)]
   | ^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `StrOptAndManualStr`
  --> tests/ui/invalid_pyclass_args.rs:98:1
   |
98 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pyclass` which comes from the expansion of the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
  --> tests/ui/invalid_pyclass_args.rs:98:1
   |
98 | #[pymethods]
   | ^^^^^^^^^^^^ multiple `__pymethod___str____` found
   |
note: candidate #1 is defined in an impl for the type `StrOptAndManualStr`
  --> tests/ui/invalid_pyclass_args.rs:89:1
   |
89 | #[pyclass(str)]
   | ^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `StrOptAndManualStr`
  --> tests/ui/invalid_pyclass_args.rs:98:1
   |
98 | #[pymethods]
   | ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0609]: no field `aaaa` on type `&Point`
   --> tests/ui/invalid_pyclass_args.rs:113:17
    |
113 | #[pyclass(str = "X: {aaaa}, Y: {y}, Z: {z}")]
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ unknown field
    |
    = note: available fields are: `x`, `y`, `z`

error[E0609]: no field `zzz` on type `&Point2`
   --> tests/ui/invalid_pyclass_args.rs:121:17
    |
121 | #[pyclass(str = "X: {x}, Y: {y}}}, Z: {zzz}")]
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unknown field
    |
    = note: available fields are: `x`, `y`, `z`

error[E0609]: no field `162543` on type `&Coord3`
   --> tests/ui/invalid_pyclass_args.rs:129:17
    |
129 | #[pyclass(str = "{0}, {162543}, {2}")]
    |                 ^^^^^^^^^^^^^^^^^^^^ unknown field
    |
    = note: available fields are: `0`, `1`, `2`
